File: 12/12.cpp
#include <iostream>
#include <cmath>
using namespace std;

unsigned long long tri_num(const int &n) {
	unsigned long long sum=0;
	for (int i=1; i <= n; i++) {
		sum += i;
	}
	return sum;
}

int count_divisors(const unsigned long long &n) {
	int count=2;
	for (auto i=2ull; i <= sqrt(n); i++) {
		if (i*i == n) count++;
		else if (n%i == 0) count+=2;
	}
	return count;
}

int main() {
	for (auto i=1ull; true; i++) {
		if(count_divisors(tri_num(i)) > 500) {
			cout << tri_num(i) << endl;
			return 0;	
		}
	}
	return 0;
}

File: 12/12.py
from math import floor, sqrt
import itertools
import time
def tri_num(n: int) -> int:
	# using the algebraic formula directly to calculate 1+2+...+n
	return n * (n+1) // 2


def count_divisors(n: int) -> int:
	count: int = 2;
	sqrt_n: int = floor(sqrt(n))
	for i in range(2, sqrt_n + 1):
		if n%i == 0: count += 2;
	
	if sqrt_n ** 2 == n: count -= 1

	return count


for num in itertools.count(1):
	if count_divisors(tri_num(num)) > 500:
		print(tri_num(num))
		break
